indexer/osm_editor.hpp

```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include "geometry/rect2d.hpp"
+
+#include "indexer/mwm_set.hpp"
+
+#include "std/function.hpp"
+#include "std/set.hpp"
+#include "std/string.hpp"
+
+struct FeatureID;
+class FeatureType;
+class Index;
+
+namespace osm
+{
+class Editor final
+{
+  struct Impl;
+  // TODO(AlexZ): Check final implementation for thread-safety.
+  Impl * m_impl;
+
+  Editor();
+  ~Editor();
+  static Impl & GetImpl();
+
+public:
+  using TEmitter = function<void(FeatureType &)>;
+  static void ForEachFeatureInRectAndScale(TEmitter f, m2::RectD const & rect, uint32_t scale);
+
+  template <class TFeaturesEmitter>
+  static void ForEachFeatureInRectAndScaleWrapper(TFeaturesEmitter & emitter,
+                                                  m2::RectD const & rect, uint32_t scale)
+  {
+    Editor::ForEachFeatureInRectAndScale(
+        [&emitter](FeatureType & ft)
+        {
+          emitter(ft);
+        },
+        rect, scale);
+  }
+
+  /// True if feature was deleted or edited by user.
+  static bool IsFeatureDeleted(FeatureID const & fid);
```

Comment:
As this method will be called from search, routing and, probably, rendering, and all these subsystems work from their own threads, this method should be thread-safe, as it interferes with DeleteFeature().